common/flatpak-dir.c

```diff
@@ -7744,6 +7748,146 @@ flatpak_dir_load_appstream_store (FlatpakDir    *self,
   return success;
 }
 
+/**
+ * flatpak_dir_get_appstream_for_ref:
+ * @self: a #FlatpakDir
+ * @remote_name: name of the remote to load the AppStream data from
+ * @ref_string: a flatpak ref to look up, in string form
+ * @arch: (nullable): name of the architecture to load the AppStream data for,
+ *    or %NULL to use the default
+ * @cancellable: (nullable): a #GCancellable, or %NULL
+ * @error: return location for a #GError
+ *
+ * Get the AppStream data for @ref_string, loading it from the locally cached
+ * AppStream XML. If the AppStream file doesn’t exist, or if @ref_string is not
+ * listed in it, %NULL will be returned without an error. An error will be
+ * returned if parsing the AppStream file fails.
+ *
+ * If @arch is %NULL, the default architecture will be used.
+ *
+ * If the @ref_string is for a runtime, %NULL will be returned immediately with
+ * no error.
+ *
+ * Returns: (transfer full) (nullable): the AppStream data for @ref_string, or %NULL
+ */
+AsApp *
+flatpak_dir_get_appstream_for_ref (FlatpakDir    *self,
+                                   const char    *remote_name,
+                                   const char    *ref_string,
+                                   const char    *arch,
+                                   GCancellable  *cancellable,
+                                   GError       **error)
+{
+  g_autoptr(FlatpakRef) ref = NULL;
+
+  ref = flatpak_ref_parse (ref_string, error);
+  if (ref == NULL)
+    return NULL;
+
+  /* Runtimes don’t have appstream. */
+  if (flatpak_ref_get_kind (ref) != FLATPAK_REF_KIND_APP)
+    return NULL;
+
+  g_autoptr(AsStore) store = as_store_new ();
+  g_autofree char *unique_id = NULL;
+
+  if (!flatpak_dir_load_appstream_store (self, remote_name, arch, store,
+                                         cancellable, error))
+    return NULL;
```

Comment:
This is a loading and fully parsing a huge compressed xml file each time, and also the source of the libappstream-glib dependency. It is also subject to freshness of the appstream deployment and divergence between the installed version and the remote HEAD version. 

The more correct way to handle this is to extract it from the per-app appdata xml file during deploy, like we do with the other appstream fields. See for example https://github.com/flatpak/flatpak/pull/2521 for an example of how the license field was added. We should do something similar for the the content rating field.

Reply:
Right, I’ll take a look at that. Thanks for the pointer.

Reply:
Done.